### PR TITLE
MOTECH-2502:Task action integration for HTTP-Agent module

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by sebbrudzinski** [comment]
test this please

File: http-agent/src/main/java/org/motechproject/http/agent/factory/HttpComponentsClientHttpRequestFactoryWithAuth.java [hunk 1]
```diff
@@ -31,14 +34,19 @@ public HttpComponentsClientHttpRequestFactoryWithAuth(String username, String pa
      */
     @Override
     protected HttpContext createHttpContext(HttpMethod httpMethod, URI uri) {
-        return createHttpContext();
+        HttpHost targetHost = new HttpHost(uri.getHost(), uri.getPort(), uri.getScheme());
```
Comment (pgesek) [inline http-agent/src/main/java/org/motechproject/http/agent/factory/HttpComponentsClientHttpRequestFactoryWithAuth.java]: We should make this configurable in the properties file, using a property called http.auth.preemptive=true/false

**Comment by sebbrudzinski** [comment]
test this please

File: http-agent/src/main/java/org/motechproject/http/agent/factory/HttpComponentsClientHttpRequestFactoryWithAuth.java [hunk 2]
```diff
@@ -21,25 +22,31 @@
...
      */
     @Override
     protected HttpContext createHttpContext(HttpMethod httpMethod, URI uri) {
-        HttpHost targetHost = new HttpHost(uri.getHost(), uri.getPort(), uri.getScheme());
-        BasicScheme basicScheme = new BasicScheme();
-        BasicAuthCache authCache = new BasicAuthCache();
-        authCache.put(targetHost, basicScheme);
         BasicHttpContext localContext = new BasicHttpContext();
-        localContext.setAttribute(ClientContext.AUTH_CACHE, authCache);
+        if (!Boolean.valueOf(settings.getProperty(HTTP_AUTH_PREEMPTIVE))) {
```
Comment (pgesek) [inline http-agent/src/main/java/org/motechproject/http/agent/factory/HttpComponentsClientHttpRequestFactoryWithAuth.java]: Shouldn't this be the other way around? If preemptive is true, then set up the cache for preemptive auth?

Comment (ghost) [inline http-agent/src/main/java/org/motechproject/http/agent/factory/HttpComponentsClientHttpRequestFactoryWithAuth.java]: I understood that if preemptive is true we will send POST without auth before one with auth. 
If you think your way is more intuitive I will change condition.

**Comment by sebbrudzinski** [comment]
test this please

**Comment by sebbrudzinski** [comment]
@pgesek ping

**Comment by sebbrudzinski** [comment]
test this please

**Comment by sebbrudzinski** [comment]
test this please

**Pull request merged**